jcl/src/java.base/share/classes/java/lang/invoke/VarHandle.java

```diff
@@ -515,6 +526,91 @@ static MethodHandle permuateHandleJEP370(MethodHandle methodHandle) {
 	}
 /*[ENDIF] Java14 */
 
+/*[IF Java15]*/
+    /**
+     * Generate a MethodHandle which translates:
+     *     FROM {VarHandle, Receiver, Intermediate ..., Value}ReturnType
+     *     TO   {Receiver, Intermediate ..., Value, VarHandle}ReturnType
+     *
+     * @param methodHandle to be permuted.
+     * @return the adapter MethodHandle which performs the translation.
+     */
+    static MethodHandle permuateHandleReferenceToJ9(MethodHandle methodHandle) {
+        /* HandleType = {Receiver, Intermediate ..., Value, VarHandle}
+         * PermuteType = {VarHandle, Receiver, Intermediate ..., Value}
+         */
+		MethodType permuteMethodType = methodHandle.type();
+		int parameterCount = permuteMethodType.parameterCount();
+		Class<?>[] params = permuteMethodType.parameterArray();
+
+		for (int i = 0; i < parameterCount; i++) {
+			if (i == 0) {
+				permuteMethodType = permuteMethodType.changeParameterType(i, params[parameterCount - 1]);
+			} else {
+				permuteMethodType = permuteMethodType.changeParameterType(i, params[i - 1]);
+			}
+		}
+
+		/**
+		 * reorder specifies the mapping between PermuteType and HandleType.
+		 * reorder = {1, 2, ..., parameterCount - 1, 0}
+		 */
+		int[] reorder = new int[parameterCount];
+		for (int i = 0; i < parameterCount; i++) {
+			if (i == (parameterCount - 1)) {
+				reorder[i] = 0;
+			} else {
+				reorder[i] = i + 1;
+			}
+		}
+
+		return MethodHandles.permuteArguments(methodHandle, permuteMethodType, reorder);
+    }
+
+	/**
+	 * Generates handleTable for Indirect VarHandles (supports JEP383).
+	 *
+	 * @param target the target VarHandle for the Indirect VarHandle.
+	 * @param handleFactory used to transform an AccessMode MethodHandle accordingly.
+	 * 
+	 * @return a MethodHandle array which is used to initialize the handleTable.
+	 */
+	static MethodHandle[] populateMHsJEP383(VarHandle target, BiFunction<AccessMode, MethodHandle, MethodHandle> handleFactory) {
+		MethodHandle[] operationMHs = new MethodHandle[AccessMode.values().length];
+
+		try {
+			MethodType replaceWithDirectType = MethodType.methodType(VarHandle.class, VarHandle.class);
+			MethodHandle replaceWithDirect = MethodHandles.publicLookup().findStatic(VarHandle.class, "asDirect", replaceWithDirectType);
+
+			for (AccessMode mode : AccessMode.values()) {
+				int index = mode.ordinal();
+				MethodHandle targetHandle = target.getMethodHandle(index);
+
+				if (targetHandle != null) {
+					MethodType targetType = targetHandle.type();
+					if (targetType.parameterType(targetType.parameterCount() - 1) == VarHandle.class) {
+						/*
+						 * Permute a J9 VarHandle to an equivalent reference implementation VarHandle
+						 * before applying handleFactory.
+						 */
+						targetHandle = permuateHandleReferenceToJ9(targetHandle);
+					}
+					operationMHs[index] = handleFactory.apply(mode, targetHandle);
+					operationMHs[index] = permuateHandleJ9ToReference(operationMHs[index]);
```

Comment:
2nd MH permutation. `Reference -> J9` conversion, in order to work with the `J9 interpreter`.